Setup Kamal for Vignettes MaMpf production deployment

Splines opened the pull request:
This PR sets up [Kamal](https://kamal-deploy.org/) such that we can deploy MaMpf for Vignettes on https://vignettes.mampf.media/. As reviewers, please read about the basics of Kamal such that terminology is clear here.

I rewrote the production Dockerfile (see `Dockerfile-vignettes`) and set up a `deploy.yml` file. It contains all the steps to deploy the Vignettes-MaMpf. This way, it should take like only a few hours to setup a completely new MaMpf server in the future (but with limitations right now, see below).

## Deploying

In Kamal, I've configured MaMpf as the main web app. This will experience zero-downtime deploys using kamal proxies. Separately managed are so-called _accessories_ like the database. They are separate Docker containers with their own lifecycle and not affected by `kamal deploy`. Instead they are managed by `kamal accessory`.

In the latest version of this PR, I've set `local: false` in the `builder` section of `deploy.yml` and configured our server to also build the image. This way, you don't have to do this on your local machine (which consumes much CPU/RAM). If you still want to, just comment out `local: false`. It will then build locally (but only if your local architecture matches amd64).

Note that Kamal only takes into account any changes that you've committed. Your commit doesn't have to be published with git push, but at least you must have a local commit. You can later still amend your commit to create a new one without having to spam 1000 Git commits when you try things out. Important: this commit restrictions appears to not apply to the Kamal accessories, there it also works without committing first. So watch out a bit.

To be able to deploy, you need to generate an SSH key for the kamal user. Then, use the following ssh config (`mampf.media` points to the same server as `vignettes.mampf.media`)

```
Host 91.98.178.89
    Hostname mampf.media
    User kamal
    IdentityFile ~/.ssh/kamal
    AddKeysToAgent yes
```

You also need the `.env` file with all the secrets. Ask for it on Basecamp.


## Limitations / Future ideas

- We use Mailtrap as SMTP relay for this setup. As we only have the free tier, emails are limited to 150/day (including error emails).
- Setting up media-forwarding correctly was not possible for me in the short time-frame. In Vignettes, you can still upload images and videos and it works fine. But not manuscripts etc. as for those we would need our /mediaforward to work correctly. I've tried setting up nginx as accessory, but as not all requests are then proxied by it, I could never get the config right for `X-Accel-Redirect`. See also [this discussion without any answers](https://github.com/basecamp/kamal/discussions/820).
- Static assets (not media like manuscripts or videos, but JS/CSS etc. assets) are served not by nginx, but by the Rails Puma web server itself. This may be a lot slower than serving over a dedicated webserver. However, setting a dedicated webserver up to work well with kamal-proxy proved difficult. See also my comments/links on the string `config.public_file_server.enabled` in `production.rb`. This is something for future PRs. It still feels blazing fast from first experiments.
- We don't use any password manager to store our secrets and pass around an `.env` file at the moment. Security-wise, this is not the best idea. But a password manager integration is only nice to have, not essential, so let's set this up in a future PR. Note that due to us using the `.env` file, all kamal command should be preprended by `bundle exec dotenv`,e .g. `bundle exec dotenv kamal deploy` instead of `kamal deploy`. Otherwise you will get missing key errors.
- It would be nice to trigger automatic builds via GitHub Actions in the future.
- There is no automatic backup solution for the database and the media mounts set up yet.

## New dependencies

- `dotenv` and `kamal`


## Off-topic

- Prometheus exporter (previously not used, but still in there as gem) is now removed since it led to problems. This is probably just some misconfiguration, but I didn't want to tackle these problems in this PR, so I removed the gem and configuration here.
- A lot of production env variables have changed, so maybe we should wait before merging this PR to calmly go through them again and adapt our old setup for "main MaMpf". Independent of whether this PR is merged or not, we can still deploy Vignettes MaMpf just by using this branch.


## ⚠ To be done for this PR

- [x] Export old database dump (probably again if something vignettes-related changed since the last dump.
- [x] Import database dump.
- [x] Get ActiveStorage secret key right. Replace by new one or better: find a solution such that this replacement doesn't have to be made in the first place.
- [x] Get ActiveStorage hardcoded url host right. Replace by new one or better: find a solution such that relative URLs are used instead of absolute ones.
- [x] Set up the error mail for vignettes, see Basecamp TODOs.
- [ ] I removed Node.js from the final build stage. However, apparently it is needed for CoffeeScript, since it uses ExecJS which in turn needs a working Node.js environment.
- [ ] Setup some better way to share the `.env` file
- [ ] Setup automatic backups

Splines committed:
Run `kamal init`

Splines committed:
Add helpful links and some more config

Splines committed:
Let AI migrate setup (for first inspiration)

Splines committed:
Create kamal user on server

Splines committed:
Add web and worker

Splines committed:
Setup ghcr registry

Splines committed:
Setup accessories

Splines committed:
Add missing accessories

Splines committed:
Add rails health check route

Splines committed:
Init env section

Splines committed:
Ignore .env file

Splines committed:
Only allow users to login via ssh keys (disable password login)

Splines committed:
Fix deploy syntax

Splines committed:
Use dotenv

Splines committed:
Add comment about docker group

Splines committed:
Fix entrypoint

Splines committed:
Add docs about folders

Splines committed:
Remove old deploy.idea.yml

Splines committed:
Fix folder permissions

Splines committed:
Don't reveal database port

Splines committed:
Fix accessory config

Splines committed:
Add missing db port secret

Splines committed:
Use DB URL

Splines committed:
Remove port

Splines committed:
Use DB_PORT variable

Splines committed:
Use PGPORT

Splines committed:
Use proper postgres variables

Splines committed:
Fix postgres env variables (docker image has different envs)

Splines committed:
Use .sh script to initialize db tables/users (fix db password issue)

Splines committed:
Fix env variable naming

Splines committed:
Init stripped-down nginx config

Splines committed:
Rnemae DB_HOST to PGHOST

see https://www.postgresql.org/docs/current/libpq-envars.html

Splines committed:
Simplify nginx deploy config

Splines committed:
Fix urls to accessories

Splines committed:
Fix nginx ports

Splines committed:
Add missing production name env variable

Splines committed:
Fix nginx config

Splines committed:
Use nginx docker image instead of openresty

Splines committed:
Move entrypoint invocation to Dockerfile

Splines committed:
Allow execution of start-mampf.sh script

Splines committed:
Change app_port to nginx container

Splines committed:
Temporarily uninstall prometheus exporter

since we get messages like these:
Prometheus Exporter, failed to send message Connection refused - connect(2) for "localhost" port 9394

Splines committed:
Try without nginx

see also: https://github.com/basecamp/kamal/discussions/1216
https://github.com/basecamp/kamal/discussions/232

Splines committed:
Don't force ssl since kamal proxy ends it

Splines committed:
Copy production to new production-new environment

Splines committed:
Set production-new env

Splines committed:
Add new production name "mampf-new"

Splines committed:
Restore old production config file

Splines committed:
Setup mail config

Splines committed:
Deploy to vignettes subdomain

Splines committed:
Don't use env to enable public file server

Splines committed:
For now, don't go to vignette subdomain

Splines committed:
Rename production name to mampf-vignette

Splines committed:
Rename production rails env (and file)

Splines committed:
Add config for production-vignette db

Splines committed:
Use snake case for rails env

Splines committed:
Indicate that production_vignette is an production environment

Splines committed:
Try to monkey-patch Rails.env.production?

Splines committed:
Fix monkey-patch

Splines committed:
Disable X-Accel-Redirect

Splines committed:
Deploy to vignettes subdomain

Splines committed:
Merge branch 'next' into kamal

Splines committed:
Setup env variables for Captcha key

Splines committed:
Merge branch 'next' into kamal

Splines committed:
Use correct RAILS_ENV for vignette deploy

Splines committed:
FIx wrong RAILS_ENV name

Splines committed:
Change error email

Splines committed:
Revert to one prod environment due to weird issues

The reason is that many code in the world depends on Rails.env.production?

We instead customize our one production environment file by using env variables.

Also use consistent naming:
"vignettes" instead of "vignette"

Splines committed:
Delete old production_vignette environment

Splines committed:
Add PRODUCTION_NAME also to dummy docker.env

Splines committed:
Don't set PGHOST, but Kamal "port" instead

Splines committed:
Hardcode port

Splines committed:
Add email authentication options

Splines committed:
Raise delivery errors in production

Splines committed:
Add url host as domain field for mails

Splines committed:
Set up SMTP config for Mailtrap

Splines committed:
Fix mail settings across env files

Splines committed:
Add shell alias

Splines committed:
Fix more mail env variables

Splines committed:
Also let project email point to new domain

Splines committed:
Add more Kamal alias

Splines committed:
Try other file locations

Splines committed:
Rework entire production Dockerfile

Splines committed:
Make spacing consistent

Splines committed:
Set correct user for /mnt

Splines committed:
Remove unused project email config

Splines committed:
Install miniracer JS V8 bridge

Splines committed:
Init media nginx server

Splines committed:
Init nginx media server configuration

Splines committed:
Make nginx config more robust

Splines committed:
Fix deploy media configs

Splines committed:
Try out docker internal host in nginx

Splines committed:
Delete unused hooks sample files

Splines committed:
Serve nginx only internally

Splines committed:
Rename nginx service

Splines committed:
Fix invalid port

Splines committed:
Remove unnecessary logs alias

(use kamal app logs instead, this even shows the full logs)

Splines committed:
Add docker internal network as extra nginx host

Splines committed:
Choose shorter name for nginx service

Splines committed:
Add link to Nginx issue with docker internal

Splines committed:
Fix wrong argument

Splines committed:
Revert changes in old production Dockerfile

Splines committed:
Install missing Shrine dependency: file

See https://github.com/shrinerb/shrine/issues/108#issuecomment-452525913

Splines committed:
Fix missing instance name key

Splines committed:
Build image on remote (and add registry caching)

see https://kamal-deploy.org/docs/configuration/builder-examples/#using-multistage-builder-cache

Splines committed:
Force build on server

Splines committed:
Fix nginx config

Splines committed:
Remove nginx accessory again

Splines committed:
Add Kamal to main section

Splines committed:
Delete nginx configurations

Splines committed:
Change to "no-reply" email address

Splines committed:
Merge branch 'next' into kamal

Splines committed:
Use mampf-hd organization for packge deployment

Splines committed:
Add comment on how DB was deployed

Splines committed:
Upgrade Pagy from v9 to v43 (#978)

Don't worry, we weren't much behind in terms of version numbers, but
Pagy made a leap:

> We needed a leap version to unequivocally signaling that it's not just
a major version: it's a complete redesign of the legacy code at all
levels, usage and API included.
> Why 43? Because it's exactly one step beyond "The answer to the
ultimate question of life, the Universe, and everything."

Not necessarily modest, but it still seems promising. And as I'd like to
use infinite scroll for the new dashboard lecture browsing, it's better
to upgrade now than later.

I've essentially followed the [upgrade
guide](https://ddnexus.github.io/pagy/guides/upgrade-guide/). There were
some major redesigns in the API, especially the fact that you don't use
`Pagy.new(...)` anymore, but instead use `include Pagy::Method` in the
`ApplicationController`, and then just call `pagy()` wherever you need
it. Also see [this
section](https://ddnexus.github.io/pagy/toolbox/paginators/#the-pagy-method).

fosterfarrell9 commented on app/workers/user_cleaner_job.rb:
```diff
@@ -15,7 +15,7 @@ def perform
     # as both mampf-experimental and mampf-next also run in production mode.
     production_name = ENV.fetch("PRODUCTION_NAME", nil)
-    return unless production_name == "mampf"
+    return if production_name != "mampf" && production_name != "mampf-vignettes"
```
For vignettes, the user cleaner should be disabled. Otherwise, study data may be destroyed as a result.

Splines replied:
Note the `unless` keyword.

Splines replied:
Oh fail, my bad, you're absolutely right. I will just revert this change I guess.

fosterfarrell9 reviewed (changes requested):
LGTM, only minor remarks. Of course, the big test will be when we deploy this when `production_name` is just `mampf`, not `mampf-vignettes`.

Splines commented on config/environments/production.rb:
```diff
@@ -81,14 +88,16 @@
   config.action_mailer.delivery_method = :smtp
   config.action_mailer.perform_deliveries = true
-  config.action_mailer.raise_delivery_errors = false
+  config.action_mailer.raise_delivery_errors = true
```
This [defaults to `true`](https://guides.rubyonrails.org/configuring.html#config-action-mailer-raise-delivery-errors). See some comments about this [here](https://stackoverflow.com/q/1325340/).

If we set it to `false`, problems like SMTP servers being down or any misconfiguration will go unnoticed by us. Of course, this means users will get a 500 page if something is wrong with email delivery, e.g. upon registration. But as a solution, either we show a nice error message in this case (and other cases where email deliver is crucial) or just show a 500 error (like this config setting does) to at least indicate something went wrong. Otherwise, errors will go unnoticed and users wonder why they get no email in their inbox.

I propose to leave it as `true` for now. If we later feel that this is too much, we can still set it to `false`.

fosterfarrell9 replied:
Fine with me.

fosterfarrell9 committed:
Fix missing slide time bug in Vignettes (#1072)

It appears that the time users spend on Vignettes slide was not
correctly transmitted to the backend after the first slide. This PR
fixes this and adds a new Playwright test for the Vignette Export.

- Note that some tests report statistic results off by 1 second, even
though we use the precise Playwright clock. For one issue, I was able to
trace this back to an implementation bug. I've noted it, but we probably
don't want to fix it now to avoid skewing the data. Another
off-by-1-second test error is unexplainable to me. But let's not waste
more energy into this, I think a variance of 1s is quite fine here.
- We couldn't get the test to run correctly in CI, so we disable it for
now. Locally it runs through fine. This is probably due to the option
`PLAYWRIGHT_HTML_OPEN=never` we use for `npx playwright test` in the CI.
We will get back to this later, but right now Müsli has priority.

## Off-topic

- Decreases playwright test timeout to 45s (was 90s beforehand)

---------

Co-authored-by: Splines <info@splines.me>
Co-authored-by: Splines <37160523+Splines@users.noreply.github.com>

Splines committed:
Update deps to those from next (see v1.20.2)

Splines committed:
Add back missing dotenv

Splines committed:
Add back missing kamal dependency

Splines commented:
We used Kamal for the media server to try many things out. But in the end, we will employ Ansible for a completely revised MaMpf setup starting with #1151. The media server will eventually be shut down. So closing this in favor of #1151.
